{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'Prepare Class I dataset as a CSV.ipynb'  class1_mhc_sequences.fasta\r\n",
      "class1_mhc_sequences.csv\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.7/site-packages/skbio/util/_testing.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as pdt\n"
     ]
    }
   ],
   "source": [
    "import skbio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>species_prefix</th>\n",
       "      <th>gene</th>\n",
       "      <th>allele</th>\n",
       "      <th>seq</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>HLA</td>\n",
       "      <td>A</td>\n",
       "      <td>01:01</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>HLA:HLA00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HLA-A*01:02</td>\n",
       "      <td>HLA</td>\n",
       "      <td>A</td>\n",
       "      <td>01:02</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFSTSVSRPGSGEPRF...</td>\n",
       "      <td>HLA:HLA00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HLA-A*01:03</td>\n",
       "      <td>HLA</td>\n",
       "      <td>A</td>\n",
       "      <td>01:03</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>HLA:HLA00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLA-A*01:04</td>\n",
       "      <td>HLA</td>\n",
       "      <td>A</td>\n",
       "      <td>01:04</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>HLA:HLA00004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HLA-A*01:06</td>\n",
       "      <td>HLA</td>\n",
       "      <td>A</td>\n",
       "      <td>01:06</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>HLA:HLA01031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14988</th>\n",
       "      <td>H-2-Kk</td>\n",
       "      <td>H-2</td>\n",
       "      <td>K</td>\n",
       "      <td>k</td>\n",
       "      <td>MAPCMLLLLLAAALAPTQTRAGPHSLRYFHTAVSRPGLGKPRFISV...</td>\n",
       "      <td>sp|P04223|HA1K_MOUSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14989</th>\n",
       "      <td>H-2-Dk</td>\n",
       "      <td>H-2</td>\n",
       "      <td>D</td>\n",
       "      <td>k</td>\n",
       "      <td>MGAMVPRTLLLLLAAALAPAQTRAGPHSLRYFETVVSRPGLGEPRF...</td>\n",
       "      <td>sp|P14426|HA13_MOUSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14990</th>\n",
       "      <td>H-2-Dp</td>\n",
       "      <td>H-2</td>\n",
       "      <td>D</td>\n",
       "      <td>p</td>\n",
       "      <td>MAPRTLLLLLAAALAPTQTRAGPHSLRYFVTAVSRPGLGKPRYMEV...</td>\n",
       "      <td>sp|P14427|HA14_MOUSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14991</th>\n",
       "      <td>H-2-Kq</td>\n",
       "      <td>H-2</td>\n",
       "      <td>K</td>\n",
       "      <td>q</td>\n",
       "      <td>PRFISVGYVDDTELVRFDSDAENPRYEPRARWMEQVEPEYWERNTQ...</td>\n",
       "      <td>sp|P14428|HA1Q_MOUSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14992</th>\n",
       "      <td>H-2-Lq</td>\n",
       "      <td>H-2</td>\n",
       "      <td>L</td>\n",
       "      <td>q</td>\n",
       "      <td>MGAMAPRTLLLLLAAALAPTQTRAGPHSMRYFETAVSRPGLGEPRY...</td>\n",
       "      <td>tr|Q31151|Q31151_MOUSE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14993 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              name species_prefix gene allele  \\\n",
       "0      HLA-A*01:01            HLA    A  01:01   \n",
       "1      HLA-A*01:02            HLA    A  01:02   \n",
       "2      HLA-A*01:03            HLA    A  01:03   \n",
       "3      HLA-A*01:04            HLA    A  01:04   \n",
       "4      HLA-A*01:06            HLA    A  01:06   \n",
       "...            ...            ...  ...    ...   \n",
       "14988       H-2-Kk            H-2    K      k   \n",
       "14989       H-2-Dk            H-2    D      k   \n",
       "14990       H-2-Dp            H-2    D      p   \n",
       "14991       H-2-Kq            H-2    K      q   \n",
       "14992       H-2-Lq            H-2    L      q   \n",
       "\n",
       "                                                     seq  \\\n",
       "0      MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "1      MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFSTSVSRPGSGEPRF...   \n",
       "2      MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "3      MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "4      MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "...                                                  ...   \n",
       "14988  MAPCMLLLLLAAALAPTQTRAGPHSLRYFHTAVSRPGLGKPRFISV...   \n",
       "14989  MGAMVPRTLLLLLAAALAPAQTRAGPHSLRYFETVVSRPGLGEPRF...   \n",
       "14990  MAPRTLLLLLAAALAPTQTRAGPHSLRYFVTAVSRPGLGKPRYMEV...   \n",
       "14991  PRFISVGYVDDTELVRFDSDAENPRYEPRARWMEQVEPEYWERNTQ...   \n",
       "14992  MGAMAPRTLLLLLAAALAPTQTRAGPHSMRYFETAVSRPGLGEPRY...   \n",
       "\n",
       "                           id  \n",
       "0                HLA:HLA00001  \n",
       "1                HLA:HLA00002  \n",
       "2                HLA:HLA00003  \n",
       "3                HLA:HLA00004  \n",
       "4                HLA:HLA01031  \n",
       "...                       ...  \n",
       "14988    sp|P04223|HA1K_MOUSE  \n",
       "14989    sp|P14426|HA13_MOUSE  \n",
       "14990    sp|P14427|HA14_MOUSE  \n",
       "14991    sp|P14428|HA1Q_MOUSE  \n",
       "14992  tr|Q31151|Q31151_MOUSE  \n",
       "\n",
       "[14993 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter, OrderedDict\n",
    "\n",
    "rows = []\n",
    "\n",
    "\n",
    "for r in skbio.read(\"class1_mhc_sequences.fasta\", format=\"fasta\"):\n",
    "    species_gene_allele = r.metadata[\"description\"].split()[0]\n",
    "    if species_gene_allele.startswith(\"H-2\"):\n",
    "        allele = species_gene_allele[-1]\n",
    "        species_and_gene = species_gene_allele[:-1]\n",
    "        gene = species_and_gene[-1]\n",
    "        species = \"H-2\"\n",
    "    else:\n",
    "        parts = species_gene_allele.split(\"*\")\n",
    "        if len(parts) == 2:\n",
    "            species_and_gene, allele = parts\n",
    "            extra = \"\"\n",
    "        else:\n",
    "            species_and_gene, extra, allele = parts\n",
    "        \n",
    "        \n",
    "        species = species_gene_allele.split(\"-\")[0]\n",
    "        gene = \"-\".join(species_and_gene.split(\"-\")[1:]) + extra\n",
    "        \n",
    "    row = OrderedDict([\n",
    "        (\"name\", species_gene_allele),\n",
    "        (\"species_prefix\", species),\n",
    "        (\"gene\", gene),\n",
    "        (\"allele\", allele),\n",
    "        (\"seq\", str(r)),\n",
    "        (\"id\", r.metadata[\"id\"]),\n",
    "    ])\n",
    "    rows.append(row)\n",
    "\n",
    "df = pd.DataFrame.from_records(rows);\n",
    "df.to_csv(\"class1_mhc_sequences.csv\", index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
